sdcm/utils/distro.py

```diff
@@ -22,30 +22,97 @@ class DistroError(Exception):
     pass
 
 
-# pylint: disable=too-many-public-methods
 @enum.unique
-class Distro(enum.Enum):
-    UNKNOWN = (None, None)
-    CENTOS7 = ("centos", "7")
-    CENTOS8 = ("centos", "8")
-    RHEL7 = ("rhel", "7")
-    RHEL8 = ("rhel", "8")
-    OEL7 = ("ol", "7")
-    OEL8 = ("ol", "8")
-    AMAZON2 = ("amzn", "2")
-    ROCKY8 = ("rocky", "8")
-    DEBIAN8 = ("debian", "8")
-    DEBIAN9 = ("debian", "9")
-    DEBIAN10 = ("debian", "10")
-    DEBIAN11 = ("debian", "11")
-    UBUNTU14 = ("ubuntu", "14.04")
-    UBUNTU16 = ("ubuntu", "16.04")
-    UBUNTU18 = ("ubuntu", "18.04")
-    UBUNTU20 = ("ubuntu", "20.04")
-    UBUNTU21 = ("ubuntu", "21.04")
-    UBUNTU21_10 = ("ubuntu", "21.10")
-    UBUNTU22 = ("ubuntu", "22.04")
-    SLES15 = ("sles", "15")
+class DistroBase(enum.Enum):
+    RHEL = enum.auto()
+    DEBIAN = enum.auto()
+    UNKNOWN = enum.auto()
+
+
+# A list of tupples of the form:
+# (enum_prefix, os name, versions list (format is either "major.minor" or "major"), boolean - debian_like == True else rehl_like)
+KNOWN_OS = [
+    ("CENTOS", "centos", ["7", "8"], DistroBase.RHEL),
+    ("RHEL", "rhel", ["7", "8"], DistroBase.RHEL),
+    ("OEL", "ol", ["7", "8"], DistroBase.RHEL),
+    ("AMAZON", "amzn", ["2"], DistroBase.RHEL),
+    ("ROCKY", "rocky", ["8"], DistroBase.RHEL),
+    ("DEBIAN", "debian", ["8", "9", "10", "11"], DistroBase.DEBIAN),
+    ("UBUNTU", "ubuntu", ["14.04", "16.04", "18.04", "20.04", "21.04", "21.10", "22.04"], DistroBase.DEBIAN),
+    ("SLES", "sles", ["15"], DistroBase.UNKNOWN),
+    ("FEDORA", "fedora", ["34", "35", "36"], DistroBase.RHEL),
+]
+
+
+enum_data = {
+}
+
+DistroBase_mapping = {
+
+}
+
+for enum_prefix, os, versions, DistroBase in KNOWN_OS:
+    if DistroBase not in DistroBase_mapping:
+        DistroBase_mapping[DistroBase] = []
+    DistroBase_mapping[DistroBase].append(os)
+    versions_by_major = {}
+    for version in versions:
+        version_major = version.split(".", maxsplit=1)[0]
+        if version_major not in versions_by_major:
+            versions_by_major[version_major] = []
+        versions_by_major[version_major].append(version)
+    for version_major, versions_with_same_major in versions_by_major.items():
+        versions_with_same_major.sort()
+        enum_data[enum_prefix + version_major] = (os, versions_with_same_major[0])
+        for version in versions_with_same_major[1:]:
+            version_for_member_names = version.replace(".", "_")
+            enum_data[enum_prefix + version_for_member_names] = (os, version)
+
+
+@static_init
+class EnumFunctionalMixin:
+    @classmethod
+    def static_init(cls):
+        def get_os_name_for_prop(name):
+            p_name = next(filter(lambda x: x[1] == name, KNOWN_OS), None)
+            if p_name is not None:
+                return p_name[0].lower()
```

Comment:
This is not safe coding, you check for `None` by then try to get the first element.
So, the condition will be `True` having empty string and the line 79 will fail in this case with `IndexError` exception.
Better to handle it correctly right away.

Reply:
It can't be absent. If we didn't return None it means that p_name has at least 2 elements (look at the filtering condition),
it checks x[1] == ... :slightly_smiling_face:

Reply:
> having empty string and the line 79

It is not a string it is a list (soon to be a tuple) from KNOWN_OS

Reply:
RIght. no problem here.